### PR TITLE
fixed ES-1396

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix for https://arangodb.atlassian.net/browse/ES-1396

* ES-1396: under some rare circumstances it was possible that background index creation missed some documents in case the documents were inserted after background index creation started and the corresponding WAL files with the inserts were already removed before background indexing caught up.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [x] Backport for 3.10: https://github.com/arangodb/arangodb/pull/17967
  - [x] Backport for 3.9: https://github.com/arangodb/arangodb/pull/17966
  - [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/17965

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/ES-1396
- [ ] Design document:

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 1]
```diff
@@ -2465,6 +2473,72 @@ RocksDBFilePurgeEnabler RocksDBEngine::startPurging() noexcept {
...
+
+  WRITE_LOCKER(lock, _walFileLock);
+
+  auto it = _lowerBoundsToKeep.find(tick);
+  if (it == _lowerBoundsToKeep.end()) {
+    _lowerBoundsToKeep.emplace(tick, 1);
+  } else {
+    // increase counter
+    ++(it->second);
+  }
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:2487]: NIT: could be simplified to this:
```suggestion
  // increase counter
  ++_lowerBoundsToKeep[tick];
```

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 2]
```diff
@@ -2465,6 +2473,72 @@ RocksDBFilePurgeEnabler RocksDBEngine::startPurging() noexcept {
   return RocksDBFilePurgeEnabler(this);
 }
 
+void RocksDBEngine::trackLowerBoundToKeep(uint64_t tick) {
+  TRI_ASSERT(tick != 0);
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:2477]: Should we make this a `ADB_PROD_ASSERT`?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 3]
```diff
@@ -2465,6 +2473,72 @@ RocksDBFilePurgeEnabler RocksDBEngine::startPurging() noexcept {
...
+  if (it == _lowerBoundsToKeep.end()) {
+    _lowerBoundsToKeep.emplace(tick, 1);
+  } else {
+    // increase counter
+    ++(it->second);
+  }
+}
+
+void RocksDBEngine::untrackLowerBoundToKeep(uint64_t tick) {
+  TRI_ASSERT(tick != 0);
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:2491]: Should we make this an `ADB_PROD_ASSERT`?

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 4]
```diff
@@ -2465,6 +2473,72 @@ RocksDBFilePurgeEnabler RocksDBEngine::startPurging() noexcept {
...
+  WRITE_LOCKER(lock, _walFileLock);
+
+  // first insert new tick. if this fails and throws, no harm is done
+  auto it = _lowerBoundsToKeep.find(newTick);
+  if (it == _lowerBoundsToKeep.end()) {
+    _lowerBoundsToKeep.emplace(newTick, 1);
+  } else {
+    // increase counter
+    ++(it->second);
+  }
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:2528]: Can be simplified to:
```suggestion
  // increase counter
  ++_lowerBoundsToKeep[tick];
```

**Review by mpoeter — APPROVED** [review]

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp [hunk 5]
```diff
@@ -376,6 +378,33 @@ arangodb::Result RocksDBBuilderIndex::fillIndexForeground() {
...
+  }
+
+  void tick(TRI_voc_tick_t tick) noexcept {
+    auto value = _tick.load(std::memory_order_acquire);
+    TRI_ASSERT(value <= tick);
+
+    // tick value must never go backwards
+    while (tick > value) {
+      if (_tick.compare_exchange_weak(value, tick, std::memory_order_release,
+                                      std::memory_order_acquire)) {
```
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp:398]: NIT
```suggestion
                                      std::memory_order_relaxed)) {
```

File: arangod/RocksDBEngine/RocksDBBuilderIndex.cpp [hunk 6]
```diff
@@ -376,6 +378,33 @@ arangodb::Result RocksDBBuilderIndex::fillIndexForeground() {
...
+ public:
+  explicit LowerBoundTracker(TRI_voc_tick_t tick = 0) noexcept : _tick{tick} {}
+
+  /// @brief earliest tick that can be released
+  [[nodiscard]] TRI_voc_tick_t tick() const noexcept final {
+    return _tick.load(std::memory_order_acquire);
+  }
+
+  void tick(TRI_voc_tick_t tick) noexcept {
+    auto value = _tick.load(std::memory_order_acquire);
```
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBBuilderIndex.cpp:392]: NIT
```suggestion
    auto value = _tick.load(std::memory_order_relaxed);
```

**Review by MBkkt — APPROVED** [review]

**Pull request merged**